libs/knowledge-store/notebooks/support.ipynb

```diff
@@ -0,0 +1,596 @@
+   ],
+   "source": [
+    "import os\n",
+    "\n",
+    "if os.getenv(\"COLAB_RELEASE_TAG\"):\n",
+    "    # (Option 1) - Set the environment variables from getpass.\n",
+    "    print(\"In colab. Using getpass/input for environment variables.\")\n",
+    "    import getpass\n",
+    "    import os\n",
+    "\n",
+    "    os.environ[\"OPENAI_API_KEY\"] = getpass.getpass(\"Enter OpenAI API Key: \")\n",
+    "    os.environ[\"ASTRA_DB_DATABASE_ID\"] = input(\"Enter Astra DB Database ID: \")\n",
+    "    os.environ[\"ASTRA_DB_APPLICATION_TOKEN\"] = getpass.getpass(\"Enter Astra DB Application Token: \")\n",
+    "\n",
+    "    keyspace = input(\"Enter Astra DB Keyspace (Empty for default): \")\n",
+    "    if keyspace:\n",
+    "        os.environ[\"ASTRA_DB_KEYSPACE\"] = keyspace\n",
+    "    else:\n",
+    "        os.environ.pop(\"ASTRA_DB_KEYSPACE\", None)\n",
+    "else:\n",
+    "    print(\"Not in colab. Loading '.env' (see 'env.template' for example)\")\n",
+    "    import dotenv\n",
+    "    dotenv.load_dotenv()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Initialize Cassio and Knowledge Store\n",
+    "With the environent variables set we initialize the Cassio library for talking to Cassandra / Astra DB.\n",
+    "We also create the `KnowledgeStore`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import cassio\n",
+    "from langchain_openai import OpenAIEmbeddings\n",
+    "from ragstack_knowledge_store import CassandraKnowledgeStore\n",
+    "from ragstack_knowledge_store.directed_edge_extractor import DirectedEdgeExtractor\n",
+    "\n",
+    "cassio.init(auto=True)\n",
+    "embeddings = OpenAIEmbeddings()\n",
+    "SITE_PREFIX=\"astra_docs\"\n",
+    "knowledge_store = CassandraKnowledgeStore(\n",
+    "    embeddings,\n",
+    "    edge_extractors = [\n",
+    "        DirectedEdgeExtractor.for_hrefs_to_urls(),\n",
+    "    ],\n",
+    "    node_table=f\"{SITE_PREFIX}_nodes\",\n",
+    "    edge_table=f\"{SITE_PREFIX}_edges\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "<cassandra.cluster.ResultSet at 0x17fb2da90>"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from cassio.config import check_resolve_session, check_resolve_keyspace\n",
+    "session = check_resolve_session()\n",
+    "keyspace = check_resolve_keyspace()\n",
+    "session.execute(f\"DROP TABLE IF EXISTS {keyspace}.{SITE_PREFIX}_nodes\")\n",
+    "session.execute(f\"DROP TABLE IF EXISTS {keyspace}.{SITE_PREFIX}_edges\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load the Documents\n",
+    "Finally, we fetch pages and write them to the knowledge store in batches of 50."
```

Comment:
Shouldn't the store handle the batching ?

Reply:
Maybe? It currently doesn't (it currently tries to traverse the whole iterable). Further, it would need to take an async iterable in the async case, which I'm not sure it does. Plan is to get this in, and then we can look at the notebook and see if we have ways to simplify it (better to have something that works / demonstrates functionality and then simplify by improving the code).

Reply:
https://github.com/datastax/ragstack-ai/blob/main/libs/knowledge-store/ragstack_knowledge_store/cassandra.py#L215

Specifically, the above will consume the iterable and try to put all the content in memory simultaneously. Agreed we should probably improve that (or at least document it). But even if it did, it wouldn't work with an `AsyncIterator`, etc. So, there is probably a case to be made that batching (and interacting with concurrency in useful ways) needs to happen outside of the client.